File: 0282_expression_add_operators/add_operators.py
from typing import List


def add_operators(num: str, target: int) -> List[str]:
    """
    Time: O(n**3)
    Space: O(n**2)
    """
    def dfs(index = 0, path = '', curr = 0, prev = None):            
        if index == len(num):
            if curr == target:
                res.append(path)
            return
        for i in range(index, len(num)):
            tmp, s = int(num[index: i + 1]), num[index: i + 1]
            if i == index or (i > index and num[index] != '0'):
                if prev is None:
                    dfs(i + 1, s, tmp, tmp)
                else:
                    dfs(i + 1, path + '+' + s, curr + tmp, tmp)
                    dfs(i + 1, path + '-' + s, curr - tmp, -tmp)
                    dfs(i + 1, path + '*' + s, curr - prev + prev * tmp, prev * tmp)
    res = []
    dfs()
    return res

File: 0282_expression_add_operators/add_operators.java
class Solution {
    // Time: O(n**3)
    // Space: O(n**2)
    public List<String> addOperators(String num, int target) {
        List<String> res = new ArrayList<>();
           StringBuilder sb = new StringBuilder();
        dfs(res, sb, num, 0, target, 0, 0);
        return res;
    }

    public void dfs(List<String> res, StringBuilder sb, String num, int pos, int target, long prev, long multi) { 
        if (pos == num.length()) {
            if (target == prev) {
                res.add(sb.toString());
            }
            return;
        }
        for (int i = pos; i < num.length(); i++) {
            if (startsWithZero(num, i, pos)) {
                break;
            }
            long curr = Long.parseLong(num.substring(pos, i + 1));
            int len = sb.length();
            if (pos == 0) {
                dfs(res, sb.append(curr), num, i + 1, target, curr, curr); 
                sb.setLength(len);
            } else {
                dfs(res, sb.append("+").append(curr), num, i + 1, target, prev + curr, curr); 
                sb.setLength(len);
                dfs(res, sb.append("-").append(curr), num, i + 1, target, prev - curr, -curr); 
                sb.setLength(len);
                dfs(res, sb.append("*").append(curr), num, i + 1, target, prev - multi + multi * curr, multi * curr); 
                sb.setLength(len);
            }
        }
    }
        private boolean startsWithZero(String num, int start, int curr) {
            return num.charAt(curr) == '0' && start != curr;
        }
    }